Remove xregexp (polyfill) & replace urlify (Django port) vendor library

lb- opened the pull request:
With #11161 merged & #11157 coming (modeladmin removal), we can now start to deprecate the global `window.URLify` and remove the legacy Regex polyfill (62KB of JS no longer needed).


## Overview

- Move vendor URLify to a TypeScript port of Django's URLify function.
- Add legacy support for window.URLify with allowUnicode as a param.
- Add release notes details for both `window.URLify` (deprecated) & `window.XRegExp` (removed).
- Update JSDoc to add more context to the port and also to the deprecated function.
- Gets us a step closer to finishing #2936 (Stop importing _editor_js.html everywhere).

## Implementation details

## About URLify `allow_unicode`

The Slug widget usage of the urlify does **not** use the Django port URLify `allow_unicode`, but it turns out this is a bug. See https://github.com/wagtail/wagtail/pull/11168#discussion_r1542436595

I have **NOT fixed this bug** in this PR but instead will raise a new bug and a new PR, as we may want to backport the fix to previous stable branches.

https://github.com/wagtail/wagtail/issues/11828


### About 'stopwords'

Our comments for `urlify` indicated that this function removes stopwords, this is not true and as far as I can tell has not been the case for many years. It only cleans the string and replaces unicode like characters with their mapped equivalent (e.g. `'À' -> 'A'`).

See https://code.djangoproject.com/ticket/11157 & https://github.com/wagtail/wagtail/pull/6410 & https://github.com/wagtail/wagtail/pull/4884

As such, I have removed this reference in the comments.

## Testing

- Check that `window.URLify` is a valid global function & accepts three args.
- Check that the page edit (new and existing pages) correctly changes the slug field when it's interacted with or changed when in sync and NOT published.
- See detailed testing for slugs on https://github.com/wagtail/wagtail/pull/10568 if needed.

<img width="1730" alt="Screenshot 2023-11-01 at 8 40 11 pm" src="https://github.com/wagtail/wagtail/assets/1396140/da0b6737-8aa3-4990-b02f-851c996e0f5b">


## Checks

-   [X] Do the tests still pass?
-   [X] Does the code comply with the style guide?
-   [X] For front-end changes: Did you test on all of Wagtail’s supported environments?
-   [X] For new features: Has the documentation been updated accordingly?

lb- commented:
Need to rebase.

lb- commented:
Ci failure appears to be unrelated to this change.

lb- commented:
Just rebased

lb- commented:
Rebased

lb- commented:
Note to self. Move release notes to 6.1

laymonage commented on wagtail/admin/static_src/wagtailadmin/js/vendor/urlify.js:
```diff
@@ -1,182 +0,0 @@
-        if (allowUnicode) {
-            // Keep Unicode letters including both lowercase and uppercase
-            // characters, whitespace, and dash; remove other characters.
-            s = XRegExp.replace(s, XRegExp('[^-_\\p{L}\\p{N}\\s]', 'g'), '');
-        } else {
-            s = s.replace(/[^-\w\s]/g, ''); // remove unneeded chars
```
How about instead of replacing this with the npm package, we only remove the `XRegExp` usage and turn this module into a util that exports urlify as a function, to be used by us via imports? Then, we expose something like `window.URLify = urlify` in `core.js` for the deprecation process? Therefore, we can still remove this file from being included as a JS vendor file in `_editor_js.html`.

I'm not strongly opposed to the NPM package, but it seems there have been some minor updates to the function in Django since the package was last released: https://github.com/django/django/commits/main/django/contrib/admin/static/admin/js/urlify.js

From a quick look they seem to be mostly cosmetic changes, but I also noticed that the package is not compatible with strict mode (see https://github.com/fyalavuz/node-parameterize/pull/11). Not sure how that affects us, but I think it's nicer to have complete control over the code in case we ever need to make changes in the future (e.g. adding more character maps).

lb- replied:
As per comments below, I did look at this approach also and decided against it. I'll take another look though. 

I also manually compared the npm package with Django's implementation (after formatting both) and they are the same. Only difference is in the custom implementation of XRegExp.

However, I get that this may be seen as too risky to use an an external dependency.

laymonage replied:
OK, no worries. I did compare them and from a quick look they seem very similar, but not exactly the same. For example, Django and us do

https://github.com/wagtail/wagtail/blob/e425244fd8a225dda2ca840d8dd3fa0b03d95243/wagtail/admin/static_src/wagtailadmin/js/vendor/urlify.js#L173

but the `parameterize` package does:

https://github.com/fyalavuz/node-parameterize/blob/7e102260ba5d107ac5d615c88e780c5eedcccf9c/parameterize.js#L191

```js
    s = s.replace(/[^-\w\s\u4E00-\u9FA5]/g, '');  // remove unneeded chars
```

The Django/our implementation trims any trailing hyphens:

https://github.com/wagtail/wagtail/blob/e425244fd8a225dda2ca840d8dd3fa0b03d95243/wagtail/admin/static_src/wagtailadmin/js/vendor/urlify.js#L178

and the `parameterize` package does not:

<img width="245" alt="image" src="https://github.com/wagtail/wagtail/assets/6379424/ffb9efe4-1ec7-4c8f-878f-613c2a53e00f">

Also, the Django implementation does `.toLowerCase()` before doing any replacements. This was done as of https://github.com/django/django/pull/12237, and I can see that our `allowUnicode` handling of `window.URLify` fallback does it this way too. Not sure what's happening but it seems the parameterize package still handles the İstanbul case correctly anyway, which is good, but it might be something to watch out for.

If you think all these differences are OK (maybe with a few tweaks e.g. trimming trailing hyphens), then I guess we could go ahead with the package.

lb- replied:
Thanks for the investigation, I'll give this another crack next week. I wish Django extracted this stuff to an NPM package, would be much easier.

Maybe we just make our own package?

We have the `@wagtail/` namespace. 

I guess at that point we can just copy it all into the code anyway.

laymonage reviewed (commented):
Thanks! I'd suggest that we keep the urlify.js instead of the NPM package, but as an exported util in our codebase. Not sure if you've explored this option yet, but let me know if you have concerns over that approach.

lb- commented:
@laymonage I did explore keeping the urlify.js but that means we also have to keep the regexp implementation it uses I think, which also relies on jQuery. 

I can go through this again but I did try to find a way to use Django's code also but it still has an old custom implementation of regex.

Hence why I found a direct port of it to modern -ish JS on npm. 

Would you be ok if I copied their implication into our package code, use it directly that way? I can add a reference to the top of the file but this is basically the same as using it via an NPM package but maybe it's easier to maintain.

lb- commented on docs/releases/6.1.md:
````diff
@@ -170,3 +170,38 @@ from wagtail import hooks
+    ]
+```
+
+If you require this for custom JavaScript functionality, it's recommended you either include your own implementation from the original [Django URLify source](https://raw.githubusercontent.com/wagtail/wagtail/main/wagtail/admin/static_src/wagtailadmin/js/vendor/urlify.js). Alternatively the [`slugify`](https://www.npmjs.com/package/slugify) or [`parameterize`, a Django URLify port](https://www.npm ...
````
Good call.

lb- commented:
Had another look and I really think we should avoid putting all this code in our codebase, the `parameterize` package is a direct copy of what Django has in it's code.

Happy to do this if you really think we should @laymonage but my understanding is that it's much easier and simpler to maintain if that code is brought in as a package.

There will be over 200 lines of code that will continue to be 'our' responsibility to think about in our code, potentially even convert to TypeScript. We already have unit tests covering the 'result' of this package in our code so I just do not see the point.

Sorry to be a pain, thank you for your review, but just thought I would double check we really really want this in our own vendor code still.

lb- commented:
@laymonage ok - you have good points, thanks for helping me understand better some of the nuanced differences and working with me on this.

I have now moved the urlify code to our own area of code instead of the external package.

Have a look and hopefully this is what you were thinking :) 

I will update the description once you have had a chance to review, explaining the new changes.

**Did some smoke testing also**

<img width="603" alt="Screenshot 2024-03-28 at 6 19 37 am" src="https://github.com/wagtail/wagtail/assets/1396140/67d50443-17c0-4974-b0eb-575612ffdcfe">
<img width="832" alt="Screenshot 2024-03-28 at 6 19 40 am" src="https://github.com/wagtail/wagtail/assets/1396140/18a30979-b365-49f6-853f-e338e627779a">

laymonage commented on client/src/entrypoints/admin/core.js:
```diff
@@ -5,6 +5,7 @@ import { Icon, Portal } from '../..';
 import { coreControllerDefinitions } from '../../controllers';
 import { escapeHtml } from '../../utils/text';
 import { initStimulus } from '../../includes/initStimulus';
+import { urlify } from '../../utils/urlify';
```
```suggestion
import { URLify } from '../../utils/urlify';
```

laymonage commented on client/src/entrypoints/admin/core.js:
```diff
@@ -32,6 +33,29 @@ window.wagtail = wagtail;
+ * @param {boolean} allowUnicode
+ * @returns {string}
+ */
+window.URLify = (str, numChars, allowUnicode) =>
+  allowUnicode
+    ? str
+        .toLowerCase()
+        .replace(/\[^-_\p{L}\p{N}\s]/g, '')
+        .replace(/^\s+|\s+$/g, '')
+        .replace(/[-\s]+/g, '-')
+        .substring(0, numChars)
+        .replace(/-+$/g, '')
+    : urlify(str, numChars);
```
We can reuse the implementation as-is, can't we?

```suggestion
window.URLify = URLify;
```

lb- replied:
yep thanks, I missed this.

laymonage commented on client/src/utils/urlify.ts:
```diff
@@ -1,21 +1,580 @@
+const downcoder = new Downcoder();
+
+/**
+ * IMPORTANT
+ * This util and the mapping is a direct port of Django's urlify.js util, without the need for a full Regex implementation.
+ * @see https://github.com/django/django/blob/main/django/contrib/admin/static/admin/js/urlify.js
+ */
+function URLify(str, numChars, allowUnicode = false) {
+  let s = str;
+  // changes, e.g., "Petty theft" to "petty-theft"
+  if (!allowUnicode) {
+    s = downcoder.downcode(s);
+  }
+  s = s.toLowerCase(); // convert to lowercase
+  // if downcode doesn't hit, the char will be stripped here
+  if (allowUnicode) {
+    // Keep Unicode letters including both lowercase and uppercase
+    // characters, whitespace, and dash; remove other characters.
+    // eslint-disable-next-line prefer-regex-literals
+    s = s.replace(s, RegExp('[^-_\\p{L}\\p{N}\\s]', 'g'), '');
+  } else {
+    s = s.replace(/[^-\w\s]/g, ''); // remove unneeded chars
   }
+  s = s.replace(/^\s+|\s+$/g, ''); // trim leading/trailing spaces
+  s = s.replace(/[-\s]+/g, '-'); // convert spaces to hyphens
+  s = s.substring(0, numChars); // trim to first num_chars chars
+  s = s.replace(/-+$/g, ''); // trim any trailing hyphens
+  return s;
 }
 
 /**
  * Returns the supplied string as a slug suitable for a URL using the vendor URLify util.
  * If the vendor util returns an empty string it will fall back to the slugify method.
  */
 export const urlify = (value: string, options = {}) => {
-  // URLify performs extra processing on the string (e.g. removing stopwords) and is more suitable
+  // URLify performs extra processing on the string and is more suitable
   // for creating a slug from the title, rather than sanitising a slug entered manually
-  const cleaned = window.URLify(value, 255);
+  const cleaned = URLify(value, 255);
```
I'm surprised we never pass the options' `unicodeSlugsEnabled` value to `URLify` here. Doesn't this mean the `WAGTAIL_ALLOW_UNICODE_SLUGS` option require you to manually edit the slug to use unicode, and that the autogenerated one (from the title) will always use ASCII? Was that intentional?

Edit: I'm not sure why/how this was changed, but, compare the following:

- https://static-wagtail-v3-0.netlify.app/admin/pages/add/base/standardpage/60/
- https://static-wagtail-v4-0.netlify.app/admin/pages/add/base/standardpage/60/

In v3.0, if you enter e.g. "Spam & ıçüş" in the title field, the slug is generated as "spam-ıçüş". In v4.0, it becomes "spam-icus". I couldn't find whether it was intentional or not, will try to dig more...

P.s. I know that historically this used to be configured via `window.unicodeSlugsEnabled` in `_editor_js.html`, and we did away with in #10189. However, that happened in 5.0, so this issue has existed since before then.

lb- replied:
Yeah, I ran into a similar confusion a long time ago when doing some earlier work (even before Stimulus on this).

I do think we should review this current behaviour but for now I will leave as is and we can raise a new bug. Any changes to the output here could cause issues for pages that are NOT published with existing slugs. This is because it will do a comparison on load of the page to check if it should keep the title/slug in sync.

lb- replied:
OK it looks like it was me sadly ->  510a6bf350e4124193e63b076398010a94e019d8 #8748 (Released in Wagtail 4.0).

The original code had.

```js
const cleaned = URLify(val, 255, window.unicodeSlugsEnabled);
```

But when moved to `client/src/utils/text.ts`, it became

```js
    const cleaned = window.URLify(val, 255);
```

I will raise a separate bug for this.

lb- replied:
Bug raised https://github.com/wagtail/wagtail/issues/11828

lb- commented:
Thanks, I'll give this another go next week.

lb- commented on client/src/utils/urlify.ts:
```diff
@@ -1,21 +1,580 @@
+  if (allowUnicode) {
+    // Keep Unicode letters including both lowercase and uppercase
+    // characters, whitespace, and dash; remove other characters.
+    // eslint-disable-next-line prefer-regex-literals
+    s = s.replace(s, RegExp('[^-_\\p{L}\\p{N}\\s]', 'g'), '');
```
Thanks I have a fix coming for this, I will resolve this for now.

lb- commented:
@laymonage  let's give this another go.

I have taken a brand new look at this as a whole and made some further refinements.

1. Pulled out the long sets of character mapping to JSON, this will be easier to maintain and understand.
2. Split up the commits to be more granular.
3. Removed our `urlify` wrapper of the Django `URLify` port and kept the urlify code to only be the urlify part, the slugify 'fallback' will now happen only in the `SlugController` code. This keeps the functions simpler, doing ONE thing only.
4. In our port of the Django urlify I have opted for (as suggested) a much simpler set of code, no classes and avoiding the linting issues. Overall it's much simpler to follow & I hope this is OK.


As noted, there is a bug that has been introduced historically in 4.0, I will not fix this in this PR and instead look at it in isolation.

I hope this makes sense, if you feel this is too much change and we should rethink this completely, I am happy to close this PR and just address this later. Hopefully though this is closer to what we want and worth the benefit of a much more maintainable/testable code with a smaller bundle size.

laymonage commented on client/src/entrypoints/admin/core.js:
```diff
@@ -32,6 +35,21 @@ window.wagtail = wagtail;
+ * @param {boolean} allowUnicode
+ * @returns {string}
+ */
+window.URLify = (str, numChars = 255, allowUnicode = false) =>
+  urlify(str, { numChars, allowUnicode }) || slugify(str, { allowUnicode });
```
Just want to note that the fallback to `slugify` wasn't in the original behaviour. I don't think it hurts to add it here, though, so I'm happy to leave it as-is.

lb- replied:
ahh good call, this is for the deprecation function so probably not a huge deal.

Thanks for the approval, I will remove this when merging in.

laymonage reviewed (approved):
Tested on Chrome 123, Firefox 125, and Safari 17.4.1 on macOS 14.4.1 and I can confirm this works well. Thanks for taking care of this and doing the TypeScript conversion!
